### DISCUSSION TITLE
At what stage should I convert input values (eg "checked" -> true)

### DISCUSSION THREAD

**Discussion by msageryd** [react-hook-form/react-hook-form#7564] [Q&A] [answered]
I'm registering all my fields when the form starts.
I'm on React-Native, so I'm unsure of what "controlled inputs" means to me. There is very little information on RN in the docs.

Some fields needs conversions. 
string -> number
"checked" -> true
"unchecked" -> false

It seems that the checked/unchecked could be taken care of if components are "controlled". If I understand correctly my fields are not "controlled" since I register them wit `register`. Please correct me if I have misunderstood.

At what stage should I convert my checkbox values "checked"/"unchecked"? 
I thought `setValueAs`could fit the bill, but according to the help text, this only applies to textInputs.

Should I make the conversion directly in the onChange method?
I need the conversion to take place before my watchers get hold of the new values, i.e. it's not enough to do the conversion on Submit.

**Comment by Moshyfawn** [comment]
It might be a completely different thing you are talking about _(then just ignore me)_, but by default RHF takes care of checkbox values. See [this CSB](https://codesandbox.io/s/checkbox-input-example-e79sl?file=/src/App.tsx)
> if I understand correctly my fields are not "controlled" since I register them wit register.

That's correct

**Reply by msageryd** [reply]
Thanks.
I might have misunderstood somehting, but I think that RHF takes care of checked/unchecked only for controlled components.

The component I'm using is not an ordinary checkbox.
https://callstack.github.io/react-native-paper/checkbox-android.html

- Its true/false/null values are "checked"/unchecked"/"intermediate"
- I'm calling `onChange` from the components `onPress` method.

**Comment by Moshyfawn** [comment, marked as answer]
Ah, so from what I understand _(excuse my ignorance, haven't touched RN on like a year, 'twas a Flutter year lol)_, [the Paper `<Checkbox />`](https://callstack.github.io/react-native-paper/checkbox.html) is expected to be a controlled component _(needs a value fed into it)_, so you better off using `useController`, I believe. That way you'll be able to handle the value conversion before `onChange` right in time for validation.
Here, see [this Snack](https://snack.expo.dev/@moshyfawn/indeterminate-state-controller)

**Reply by msageryd** [reply]
Thanks again.
I'll do it the way you propose.

I would love to have some api support for this kind of conversion to keep everything in the same place. To me this is a dataType conversion. I think it would fit well into `setValueAs`. @bluebill1049 What is the rationale for limiting setValueAs to only work with textInputs?

I'm leaning towards making a "middleware"-onChange which can handle dataType conversions for number, date, checkbox-strings, etc. This way I will have all my type conversions in one place.

**Reply by Moshyfawn** [reply]
It feels like I've missed something about your use-case, tbh. See [this CSB](https://codesandbox.io/s/new-https-ute28?file=/src/App.tsx). It uses HTML select element instead of a tri-state checkbox for simplicity; same idea:  3 options "checked", "unchecked", "indeterminate". Using `setValueAs`, we can convert string values to `true`, `false` and `null`:
```js
setValueAs: (value) => {
  if (value === "checked") {
    return true;
  }
  if (value === "unchecked") {
    return false;
  }
    return null;
  }
```
As it's an uncontrolled input _(register doesn't return `value`)_, select has it's own string values and we get our `boolean` or `null` on submit. Now, the first problem is setting `defaultValues` for out select _(`isAgreed` in case of [my CSB](https://codesandbox.io/s/new-https-ute28?file=/src/App.tsx:1241-1511))_ wouldn't work, as the source of truth value should be one of 3 checkbox states, but `isAgreed` should be a `boolean` or `null` _(see the visual select element state on initial load)_.
Another thing is that your Paper `<Checkbox />` doesn't return any value / event from `onPressed`; you need to `watch` the form value to be able to set a new value on `onPressed` event. Further more, there's no event to pass into the `onChange`.
Finally, your Paper `<Checkbox />` is designed to be a controlled input. You still need to `watch` for the form value change, and it's a case for `useController`. Looking at [it's RHF docs](https://react-hook-form.com/api/usecontroller/):
> Important: doesn't support valueAs for useController.

The value conversion for controlled inputs happens `onChange` that you provide with `field.onChange`.

**Reply by msageryd** [reply]
Your code example is great, and I have implemented it.

But I will need other dataType conversions for other fields. Ex:
- name: fieldA
- dataType: integer 
- component: textInput

These parameters will make my textInput component show a numerical keyboard and won't accept any decimal signs. But users can avoid my keyboard-limitation by calling up another keyboard, for example an emoji-keyboard. 

I want to convert the inputString to a rounded numerical on onChange (possibly with a 100ms delay) so the numerical value gets stored in RHF as the field value. I cannot wait until onBlur since I have other checks that I need to run against the numerical value. Trying to cast the string as number should be done in a validator so an error can be set. But the rounding to int and other types of formatting does not need to sit in a validator.

This conversion could be taken care of in onChange.
A similar conversion might be needed in the other direction when `value` is send to teh controlled component. In the integer case this won't be needed, since it will be casted automatically, but in the checkbox case a special conversion is needed.

My thoughts about this is:
- when is setValueAs supposed to be used? Seems like a candidate for this use case, but it's limited to textInputs according to the docs
- maybe better to chain the onChange-methods so I can have a centralized "middleware-onChange" where all dataType conversion takes place.
